ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,225 @@ func (ps *prioritySampler) apply(spn *span) {
 	}
 	spn.SetTag(keySamplingPriorityRate, rate)
 }
+
+type rulesSampler struct {
+	rules   []SamplingRule
```

Comment:
Can we please benchmark `StartSpan` in three situations:

* Without rule sampling
* With rule sampling
* With rule sampling where each rule is a pointer

And post the results? We are working on optimising `StartSpan` for some time now and it is relevant to check this.

Reply:
Wasn't getting consistent results via StartSpan benchmarks, so went with a direct comparison of rules sampler implementations.

Benchmark uses 3 rules with the 3rd one being the match to the span. The expectation is that users won't need a lot of rules.

Results:
```sh-session
$ go test -run=none -bench=BenchmarkRulesSamplerImpls -benchtime=10s
goos: linux
goarch: amd64
pkg: gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer
BenchmarkRulesSamplerImpls/structs-4         	36743371	       324 ns/op
BenchmarkRulesSamplerImpls/pointers-4        	36370348	       342 ns/op
PASS
ok  	gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer	25.515s
$ go test -run=none -bench=BenchmarkRulesSamplerImpls -benchtime=10s
goos: linux
goarch: amd64
pkg: gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer
BenchmarkRulesSamplerImpls/structs-4         	36933296	       324 ns/op
BenchmarkRulesSamplerImpls/pointers-4        	36172393	       322 ns/op
PASS
ok  	gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer	24.292s
$ go test -run=none -bench=BenchmarkRulesSamplerImpls -benchtime=10s
goos: linux
goarch: amd64
pkg: gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer
BenchmarkRulesSamplerImpls/structs-4         	36502827	       325 ns/op
BenchmarkRulesSamplerImpls/pointers-4        	36558997	       328 ns/op
PASS
ok  	gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer	24.544s
```

Reply:
Thanks Caleb. Can you please post the results for the other benchmark too? `no-rules` and `with-rules` etc. Would be nice to see all of them.

Reply:
```sh-session
$ go test -run=none -bench='^BenchmarkRulesSampler$' -benchtime=10s
goos: linux
goarch: amd64
pkg: gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer
BenchmarkRulesSampler/no-rules-4         	 8031397	      1773 ns/op
BenchmarkRulesSampler/unmatching-rules-4 	 7459657	      1858 ns/op
BenchmarkRulesSampler/matching-rules-4   	 7754979	      1833 ns/op
PASS
ok  	gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer	56.915s
$ go test -run=none -bench='^BenchmarkRulesSampler$' -benchtime=10s
goos: linux
goarch: amd64
pkg: gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer
BenchmarkRulesSampler/no-rules-4         	 6115885	      1714 ns/op
BenchmarkRulesSampler/unmatching-rules-4 	 8387515	      1657 ns/op
BenchmarkRulesSampler/matching-rules-4   	 6550923	      1563 ns/op
PASS
ok  	gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer	49.032s
```

Those numbers didn't make sense, so I went with the other more-direct benchmark of the part you'd asked about.